lib/kcerts/ssh.go

```diff
@@ -0,0 +1,197 @@
+
+import (
+	"crypto/rand"
+	"crypto/rsa"
+	"encoding/json"
+	"fmt"
+	"github.com/enfabrica/enkit/lib/cache"
+	"github.com/mitchellh/go-homedir"
+	"golang.org/x/crypto/ssh"
+	"io/ioutil"
+	"os"
+	"os/exec"
+	"path/filepath"
+	"regexp"
+	"strconv"
+	"strings"
+	"time"
+)
+
+const CAPrefix = "@cert-authority"
+const SSHDir = ".ssh"
+const KnownHosts = "known_hosts"
+
+const SSHCacheKey = "enkit_ssh_cache_key"
+
+// FindSSHDir will find the users ssh directory based on $HOME. If $HOME/.ssh does not exist
+// it will attempt to create it.
+func FindSSHDir() (string, error) {
+	hDir, err := homedir.Dir()
+	if err != nil {
+		return "", fmt.Errorf("could not find the home directory: %w", err)
+	}
+	sshDir := filepath.Join(hDir, SSHDir)
+	if err := os.Mkdir(sshDir, 0700); err != nil && !os.IsExist(err) {
+		return "", fmt.Errorf("could not create directory %s: %w", sshDir, err)
+	}
+	return sshDir, nil
+}
+
+// AddSSHCAToClient adds a public key to the $HOME/.ssh/known_hosts in the ssh-cert x509.1 format.
+// For each entry, it adds an additional line and does not concatenate
+func AddSSHCAToClient(publicKey ssh.PublicKey, hosts []string, sshDir string) error {
+	caPublic := string(ssh.MarshalAuthorizedKey(publicKey))
+	knownHosts := filepath.Join(sshDir, KnownHosts)
+	knownHostsFile, err := os.OpenFile(knownHosts, os.O_RDWR|os.O_CREATE|os.O_APPEND, 0644)
+	defer knownHostsFile.Close()
+	if err != nil {
+		if os.IsNotExist(err) {
+			return fmt.Errorf("could not create known_hosts file: %w", err)
+		}
+		return err
+	}
+	existingKnownHostsContent, err := ioutil.ReadAll(knownHostsFile)
+	if err != nil {
+		return fmt.Errorf("error reading %s: %w", knownHosts, err)
+	}
+	for _, dns := range hosts {
+		// caPublic terminates with a '\n', added by ssh.MarshalAuthorizedKey
+		publicFormat := fmt.Sprintf("%s %s %s", CAPrefix, dns, caPublic)
+		if strings.Contains(string(existingKnownHostsContent), publicFormat) {
+			continue
+		}
+		_, err = knownHostsFile.WriteString(publicFormat)
+		if err != nil {
+			return fmt.Errorf("could not add key %s to known_hosts file: %w", publicFormat, err)
+		}
+	}
+	return nil
+}
+
+type sshCache struct {
+	PID       int       `json:"pid"`
+	Sock      string    `json:"sock"`
+	TimeStamp time.Time `json:"time_stamp"`
+}
+
+// FindSSHAgent Will start the ssh agent in the interactive terminal if it isn't present already as an environment variable
+// Currently only outputs the env and does not persist it across terminals
+func FindSSHAgent(store cache.Store, ttl time.Duration) (string, int, error) {
+	// TODO(adam): sniff out dead agents and sneak pids
+	envSSHSock := os.Getenv("SSH_AUTH_SOCK")
+	envSSHPID := os.Getenv("SSH_AGENT_PID")
+	if envSSHSock != "" && envSSHPID != "" {
+		pid, err := strconv.Atoi(envSSHPID)
+		if err != nil {
+			return "", 0, fmt.Errorf("%s is not a valid pid: %w", envSSHPID, err)
+		}
+		return envSSHSock, pid, nil
+	}
+
+	sshEnkitCache, isFresh, err := store.Get(SSHCacheKey)
+	if err != nil {
```

Comment:
here and below: I would refactor the error handling so that the cache is entirely optional.

Eg, if we fail to write in cache because we are out of disk space, or disk is faulty, or ... or if we fail to read a cache file (corrupted, ...) we should still be able to ssh into other boxes.

Further, we have the same problem as above: the cache may be bogus. For example: started an agent, enjoyed for a while, machine was rebooted... now cache points to an agent that no longer exists.